docs/MultipleLoadNativeModule.md

```diff
@@ -0,0 +1,124 @@
+entry and load the driver.
+
+**Note**: In this option, the application can simply pass the driver name (with absolute or relative path) to libbpf /
+eBPF API and ebpfapi.dll can internally generate a unique service name.
+
+#### `GET_NATIVE_MODULE_HANDLE` Design:
+1. UM will pass the service name. ebpfcore looks for the module based on the service name.
+2. If ebpfcore does not find the module, it does the following:
+    1. Generate a GUID for the client.
+    2. Create a placeholder native module and add it to hash tables.
+    3. Populate the service path (created by appending the service name to path prefix), and the module ID in the
+    native module.
+    4. Return the module HANDLE, module ID and return code `EBPF_SERVICE_NOT_PRESENT`.
+3. If runtime finds the module, and it not in "unloading" state, it will return a new handle to module.
+4. If runtime finds the module, and it is in "placeholder" state, it will block on an event, which will be set when the
+module is loaded.
+5. If runtime finds the module, and it is in "unloading" state, it will block on another event, which will be set when
+call to driver unload returns. Once this event is set, this thread will be unblocked and move to step 2.
+
+#### Misc. Changes
+1. ebpfcore will now maintain 2 hash tables to look up the native modules. One hash table has service name as key, and
+another has module ID as key.
+2. Currently native module is deleted from hash table when ref count becomes 0, which is before the driver is unloaded.
+In this new design the native module should be deleted from the hash table only when the driver unload call has
+completed. Also, we need to ensure that driver unload happens inline.
+
+### Option 2
+Another approach is to completely skip all the above synchronization logic in ebpfcore and ebpfapi.dll, and separate
+out service install / uninstall completely from the scope of the application. Native modules will be installed by some
+installer that installs the native module on the system. Once that is done, programs can be loaded / unloaded multiple
+times, without the need of install / uninstall of service entries.
+
+Changes needed to implement this:
+1. Installer installs the service entry. We need to ensure the service name is unique. Two native modules with same
+name but in different locations should be installed as 2 different services. Installer also populates the client module
+ID registry key with a random GUID.
+2. The app that loads the program will pass the service name to ebpfapi (TBD: How does app know the service name for its
+service). We can still use the existing `bpf_object__open` API, and the API can take the service name instead of file
+name as input.
+
+**Note**:
+1. One limitation with option 2 is that all existing applications that are using native eBPF programs need to now
+also have an explicit installer / un-installer for the native programs that are they are using. This installer can
```

Comment:
I disagree with this sentence.  My feedback was that the applications using the native program do NOT install the native program.  They only load it after it was installed by the same mechanism that got the native program on the machine in the first place.  E.g. that could be L3AFD or a script or whatever else.  It's also possible that the installer of the app and the installer of the native program are the same program, but they're logically two separate things which is why this sentence makes no sense to me.  An app installer can have a dependency on another components installer, just like multiple programs depend on say the .NET Framework being installed first, but that doesn't mean that multiple programs' installers all know the details of how to install the .NET Framework.  That's not how it works.

Reply:
@Alan-Jowett as per today's discussion in the triage meeting, for option 2, if the proposal is that ebpfapi exposes APIs for installers that helps install / uninstall the drivers, does it become similar to the approach in the below draft PR that adds 2 new APIs `ebpf_initialize_native_program_state()` and `ebpf_uninitialize_native_program_state()` for creating the service?

[#3556](https://github.com/microsoft/ebpf-for-windows/pull/3556/files#diff-8b744ef020b2dae7afefac77b8d1d9664fa3d249c683b307cd393362da9f5e8b)

Reply:
Updated doc with what was discussed in the triage meeting.